server/src/ws/routes.ts

```diff
@@ -1,50 +1,117 @@
-  // Auth is probably unnecessary, because to connect to the websocket you need to be logged in on the client anyway
-  // Uncomment this line to enable authentication
-  // server.addHook('preValidation', server.authenticate!) // TODO: make TS happy with this line
-
-  server.get<{ Querystring: { username: string } }>(
+  server.get<{ Querystring: { jwt: string } }>(
     '/',
     { websocket: true },
     async (connection, req) => {
       // A user join the chat
-      // const { id } = req.user as any
-      // const userAccounts = await usersStorageClient.load('allUsers')
-      // const user = userAccounts.find((account) => account.user.id === id)
-      // const username = user ? user.name : 'Unknown user'
-
-      broadcast({
-        sender: '__server',
-        message: `${'TODO'} joined`,
-      })
+      const { jwt } = req.query
+      const userId = getUserIdFromJwt(jwt)
+      const user = await getUserFromId(userId)
+      if (user) {
+        serverBroadcast(`${user.name} joined`)
+      }
 
       // A user leaving the chat
-      connection.socket.on('close', () => {
-        broadcast({
-          sender: '__server',
-          message: `${req.query.username} left`,
-        })
+      connection.socket.on('close', async () => {
+        const { jwt } = req.query
+        const userId = getUserIdFromJwt(jwt)
+        const user = await getUserFromId(userId)
+
+        if (user) {
+          serverBroadcast(`${user.name} left`)
+        }
       })
 
       // Broadcast incoming message
-      connection.socket.on('message', (message) => {
-        message = JSON.parse(message.toString())
+      connection.socket.on('message', async (msg) => {
+        msg = JSON.parse(msg.toString())
+        const incomingMessage = msg as unknown as {
+          threadId: string
+          message: string
+        }
+
+        const { jwt } = req.query
+        const userId: UserId = getUserIdFromJwt(jwt)
+
+        const res = await chatClient.AddMessageToThread(
+          incomingMessage.threadId,
+          userId,
+          incomingMessage.message
+        )
+
+        if (res.error) {
+          serverBroadcast(res.error)
+          return
+        }
+
+        broadcast(new WsMessageEvent('add', res.thread.id, res.message))
+
+        // If user is chatting with the bot:
+        if (
+          res.thread.participants.length === 2 &&
+          res.thread.participants.includes('autofriendid')
+        ) {
+          await sleep(3000)
+          const msg = chatGptClient
+            ? await chatGptClient.getResponse(incomingMessage.message)
+            : randChoice([
+                "That sound's great",
+                'Sure thing :)',
+                'How kind of you to say',
+              ])
+          const botRes = await chatClient.AddMessageToThread(
+            res.thread,
+            'autofriendid',
+            msg
+          )
+
+          if (res.error) {
+            serverBroadcast(res.error)
```

Comment:
This should only be broadcast to the sender. Also, when the client receives it, it will have no idea how to action this item - all the client knows is "I tried to add a message to thread 123"; we need to eventually tell the client that their specific request failed, which is a bit complicated. I think incoming requests will have to have a client-generated ID that the server can reference for future comms, so the client can tie a message like 

{
  type: 'server-error',
  clientId: 'c120293'
 }
 
 The client that sent the original message and generated that id will then be able to do something with the response (like show an error icon next to the message that failed, which can then be retried)

Reply:
Make sense. Will try to implement it if there's time